src/common/tracer.cc

```diff
@@ -73,6 +73,41 @@ bool Tracer::is_enabled() const {
   return g_ceph_context->_conf->jaeger_tracing_enable;
 }
 
+void encode(const jspan_context& span_ctx, bufferlist& bl, uint64_t f) {
+  ENCODE_START(1, 1, bl);
+  using namespace opentelemetry;
+  using namespace trace;
+  auto is_valid = span_ctx.IsValid();
+  encode(is_valid, bl);
+  if (is_valid) {
+    encode_nohead(std::string_view(reinterpret_cast<const char*>(span_ctx.trace_id().Id().data()), TraceId::kSize), bl);
```

Comment:
After some drilling in the encoding stuff, I started worrying there is (and was; not a new thing) a rather minor problem with the tracker's format. By using `encode_nohead` an encoder enforces decoders to use their own, local instance of the `TraceId::kSize` and `SpanId::kSize` constants.

The current assumption is they match between e.g. revisions of `jspan_context` and architectures. If this invariant breaks out, we could run into:

1. if decoder's constant `<` encoder's constant, then decoder's compat-preserving machinery will silently eat the extra bytes (`bl += struct_end - bl.get_off()`) and what happen later is unknown:
```cpp
/**
 * finish decode block
 *
 * @param bl bufferlist::iterator we were decoding from
 */
#define DECODE_FINISH(bl)                                               \
  } while (false);                                                      \
  if (struct_end) {                                                     \
    if (bl.get_off() > struct_end)                                      \
      throw ::ceph::buffer::malformed_input(DECODE_ERR_PAST(__PRETTY_FUNCTION__)); \
    if (bl.get_off() < struct_end)                                      \
      bl += struct_end - bl.get_off();                                  \
  }

```
2. if decoder's constant `>` encoder's constant – `end_of_buffer` or `malformed_input` will be thrown.

I doubt it's a severe defect as it can happen only on the tracing (`is_valid`) path but if we consider a feature bit, can we fix this minor as well? Is it worth? Is a note in the doc an alternative?

Reply:
thanks @rzarzynski, this issue was raised by @idryomov here: https://github.com/ceph/ceph/pull/51043
will look at that once the end2end work is re-merged